### ISSUE TITLE
Standard creation functions for spatial graphs from only points

### ISSUE THREAD

**Issue by luukvdmeer** [luukvdmeer/sfnetworks#52]
**Is your feature request related to a problem? Please describe.**
You can now create an `sfnetwork` from only a set of spatial linestrings. The linestrings will be the edges, and the endpoints of the linestrings the nodes. Also, you can create an `sfnetwork` from only a set of spatial points. The points will be the nodes, and edges will be created in a sequential order (i.e. from node 1 to 2, node 2 to 3, etc). This was created as a solution for this [StackOverflow issue](https://stackoverflow.com/questions/57833905/how-to-convert-a-list-of-sf-spatial-points-into-a-routable-graph).

However, there are much more ways to create a spatial graph from only a set of points. [Dale (2017)](https://www.cambridge.org/core/books/applying-graph-theory-in-ecological-research/spatial-graphs/0D50525C2A011B60A796446D8F153BC7) lists the most common ones:

![sn](https://user-images.githubusercontent.com/26540305/83967281-47cc9b00-a8c0-11ea-9f3a-af2b46aa2013.png)

**Describe the solution you'd like**
More options to create an `sfnetwork` from only a set of spatial points. This could be implemented with an extra argument to `as_sfnetwork.sf`, specifying which creation algorithm to use, and also in the same style as tidygraphs `create_*` family of functions. That is: `create_mnn`, `create_nn`, `create_mst`, et cetera.

**Comment by JosiahParry** [comment]
regarding minimum spanning tree this can be accomplished using spdep and sfdep. here's an example of creating an sfnetwork using an mstree. 

``` r
library(sf)
library(dplyr)
library(sfdep)
library(sfnetworks)

geo <- sfdep::guerry_nb |>
  select(geometry) |>
  mutate(nb = st_complete_nb(n()),
         wt = st_nb_dists(geometry, nb))
#> ! Polygon provided. Using point on surface.

listw <- recreate_listw(geo$nb, structure(geo$wt,  "B" = TRUE))

ms <- spdep::mstree(listw)

ms_df <- as.data.frame(ms) |>
  rlang::set_names(c("from", "to", "edge_length"))


geom <- sf::st_centroid(geo)[["geometry"]]
#> Warning in st_centroid.sf(geo): st_centroid assumes attributes are constant over
#> geometries of x
nodes <- geom

from <- geom[ms_df$from]
to <- geom[ms_df$to]

from <- as_tibble(st_coordinates(from)) |>
  mutate(id = row_number())
to <- as_tibble(st_coordinates(to)) |>
  mutate(id = row_number())



# need to find a way to arrange using base R
edges2 <- sfheaders::sf_linestring(
  arrange(bind_rows(from, to), id),
  x = "X",
  y = "Y",
  linestring_id = "id") |>
  bind_cols(ms_df)

plot(sfnetwork(nodes, edges2))
#> Checking if spatial network structure is valid...
#> Spatial network structure is valid
```

![](https://i.imgur.com/dbW672x.png)

<sup>Created on 2022-08-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by JosiahParry** [comment]
Here is also a proof of concept for creating sfnetwork based on delaunay triangulation using {geos}. The challenge with this approach is getting the attributes associated back to the initial points. I suspect a spatial join could accomplish it?

``` r
library(sf)
library(dplyr)
library(sfnetworks)

nodes <- st_centroid(sfdep::guerry)
#> Warning in st_centroid.sf(sfdep::guerry): st_centroid assumes attributes are
#> constant over geometries of x

es <- geos::geos_make_collection(nodes) |>
  geos::geos_delaunay_edges() |>
  geos::geos_unnest(keep_multi = FALSE)


# cast to sfnetwork
sfn <- es |>
  st_as_sf() |>
  as_sfnetwork()

plot(sfn)
```

![](https://i.imgur.com/s0g7aVi.png)

``` r

# an idea as how to get the attributes back to the points
nodes |> st_as_sf() |>
  mutate(id = row_number()) |>
  st_join(sfn |>
            st_as_sf() |>
            mutate(id2 = row_number())) |>
  select(id, id2)
#> Simple feature collection with 85 features and 2 fields
#> Geometry type: POINT
#> Dimension:     XY
#> Bounding box:  xmin: 125581.7 ymin: 1732960 xmax: 983954.9 ymax: 2610980
#> CRS:           NA
#> # A tibble: 85 × 3
#>       id   id2           geometry
#>    <int> <int>            <POINT>
#>  1     1    75 (832852.3 2126601)
#>  2     2    15 (688485.6 2507622)
#>  3     3    84 (665510.1 2155203)
#>  4     4    30 (912995.8 1908303)
#>  5     5    31 (911433.9 1970312)
#>  6     6    54 (765421.3 1974521)
#>  7     7    14 (766676.6 2515627)
#>  8     8    25 (531866.4 1769020)
#>  9     9    35 (735432.7 2368847)
#> 10    10    26   (606310 1788959)
#> # … with 75 more rows
#> # ℹ Use `print(n = ...)` to see more rows
```

<sup>Created on 2022-08-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by JosiahParry** [comment]
Creating an sfnetwork using spdep + sfdep is rather straightforward as well. 

``` r
library(sf)
library(dplyr)
library(spdep)
library(sfnetworks)

nodes <- st_centroid(sfdep::guerry)
#> Warning in st_centroid.sf(sfdep::guerry): st_centroid assumes attributes are
#> constant over geometries of x

geo <- st_geometry(nodes)
nb <- tri2nb(geo)

plot(nb, geo)
```

![](https://i.imgur.com/tBoGtSG.png)

``` r

wt <- sfdep::st_nb_dists(geo, nb)

sfdep::st_as_graph(geo, nb, wt) |> 
  plot()
#> Checking if spatial network structure is valid...
#> Spatial network structure is valid
```

![](https://i.imgur.com/TztWOzl.png)

<sup>Created on 2022-08-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by luukvdmeer** [comment]
This is now implemented in v1.0, see https://luukvdmeer.github.io/sfnetworks/articles/sfn02_create_represent.html#from-spatial-points
